### ISSUE TITLE
Dimension transposed after tensordot

### ISSUE THREAD

**Issue by ydup** [numpy/numpy#12111]
Hello,
I try to product a matrix(2D) with another one (3D), but find that the dimension of the result is transposed. 
```python
diagMat=np.diag([1,1,1])
mat3D = np.array([[[ 0,  1,  2, 10],
        [ 3,  4,  5, 0],
        [ 6,  7,  8, 1]],
       [[ 9, 10, 11, 2],
        [12, 13, 14,3],
        [15, 16, 17,0]]])
newMat = np.tensordot(mat3D, diagMat, axes=([1],[1]))
print mat3D
print newMat
print newMat.shape
# --> (2, 4, 3)
print mat3D.shape
# --> (2, 3, 4)
```
When the specific dimension changes, I don't want to see that the other dim transposes. By the way, because I am going to calculate high dimension matrix with tensordot, it can be tedious for me to re-transpose them. Such as manully transpose the dim-1 and dim-2 of the ```newMat```.
```python
import numpy as np
diagMat=np.diag([1,1,1])
mat3D = np.array([[[ 0,  1,  2, 10],
        [ 3,  4,  5, 0],
        [ 6,  7,  8, 1]],
       [[ 9, 10, 11, 2],
        [12, 13, 14,3],
        [15, 16, 17,0]]])
newMat = np.tensordot(mat3D, diagMat, axes=([1],[1]))
print mat3D
print newMat.swapaxes(1,2)
print newMat.swapaxes(1,2).shape
print mat3D.shape
```
Hope that this description is clear for you : )

**Comment by ydup** [comment]
For the product of tensor and 2D matrix, I pull a request on https://github.com/numpy/numpy/pull/12114. Please feel free to test it : )

**Comment by shoyer** [comment]
Your best option for full control over the order of axes from `tensordot` is to use `np.einsum` instead. We are unlikely to add `tensordotmat` since it's a rather specialized use-case and there are easy work-arounds.

**Comment by ydup** [comment]
Thanks! @shoyer 
With your advise, I finish this with ```np.einsum``` as follow:
```python
np.einsum('aib,ki->akb', mat3D, diagMat)
```
But I still have a question. Is that the order of dimension changes using tensordot a bug? At the source code of ```tensordot```, we can see that it returns to a matrix reshaped into ```(olda+oldb)```. It can be tedious for us to finish the dot production if it exists shift of dim. So I fix it with the rules of shift to original dim order based on ```tensordot```, but I am just sure that it can work in the tensor dot 2D matrix, then I pull the request in 12114 with the ```tensordotmat```. 
I wonder if this problem in the ```tensordot``` can be solved.

**Comment by shoyer** [comment]
Oddly, `np.tensordot` does not document the shape of its return value in its docstring. This should be fixed.

I do think it is following a reasonable rule for the axes order: all axes from the first argument appear, followed by all axes from the second argument. Even if it wasn't the most useful rule, we would need to be quite cautious with changing it because it would assuredly break user code.

**Comment by ydup** [comment]
Yes, according to the algorithm of ```tensordot```, the new axes order of the output is based on the calculation order.
Worrying about that the adding rules to ```tensordot``` will affect the previous usage of ```tensordot```, I think that writing a new one function with fixing the axes shift and pointing out the difference between the new one and ```tensordot``` can be a soft way for users.

**Comment by eric-wieser** [comment]
Simplifying:
```
np.einsum('aib,ki->akb', mat3D, diagMat)
np.einsum('...ib,...ki->...kb', mat3D, diagMat)
np.einsum('...ki,...ib->...kb', diagMat, mat3D)
diagMat @ mat3D
```

Your proposed function already exists, and it's `matmul` / `@`

**Comment by ydup** [comment]
@eric-jones Yes, definitely, you get a simplified way to solve that the 3D matrix dot product with a 2D one. But this issue aims at the problem existing in the ```tensordot```. In another word, the problem of axes change in the ```tensordot``` mismatches the usual mathematical analysis of tensor dot product. And this could cause the tedious work for multiple tensor dot product.

**Comment by eric-wieser** [comment]
So using `einsum`-like notation, you're looking for `xxxjxxx,jk->xxxkxxx`, where `x` is some non-contracted dimension?

**Comment by ydup** [comment]
Yes. And my purpose of writing the ```tensordotmat``` is to fix the problem of the non-contracted dimensions change. In another word, I mean ```tensordotmat``` could be a soft way to fix the issue of ```tensordot```.

**Comment by mattip** [comment]
Closing, there is a work-around using tensordot and specifying the order.
